crates/sui-indexer-alt-jsonrpc/src/data/kv_loader.rs

```diff
@@ -41,12 +49,62 @@ pub(crate) enum TransactionContents {
 }
 
 impl KvLoader {
-    pub(crate) fn new_with_bigtable(bigtable_loader: Arc<DataLoader<BigtableReader>>) -> Self {
-        Self::Bigtable(bigtable_loader)
+    pub(crate) fn new_with_bigtable(
+        bigtable_loader: Arc<DataLoader<BigtableReader>>,
+        metrics: Arc<RpcMetrics>,
+        config: Arc<RpcConfig>,
+    ) -> Self {
+        Self::Bigtable(bigtable_loader, metrics, config)
     }
 
-    pub(crate) fn new_with_pg(pg_loader: Arc<DataLoader<PgReader>>) -> Self {
-        Self::Pg(pg_loader)
+    pub(crate) fn new_with_pg(
+        pg_loader: Arc<DataLoader<PgReader>>,
+        metrics: Arc<RpcMetrics>,
+        config: Arc<RpcConfig>,
+    ) -> Self {
+        Self::Pg(pg_loader, metrics, config)
+    }
+
+    fn metrics(&self) -> &Arc<RpcMetrics> {
+        match self {
+            Self::Bigtable(_, metrics, _) => metrics,
+            Self::Pg(_, metrics, _) => metrics,
+        }
+    }
+
+    fn config(&self) -> &Arc<RpcConfig> {
+        match self {
+            Self::Bigtable(_, _, config) => config,
+            Self::Pg(_, _, config) => config,
+        }
+    }
+
+    async fn measure_load<T, E>(
+        &self,
+        load: impl Future<Output = Result<T, E>>,
+        load_desc: String,
+    ) -> Result<T, E> {
+        let metrics = self.metrics();
+        let threshold_seconds = self.config().slow_request_threshold.as_secs() as f64;
+        metrics.kv_requests_received.inc();
+
+        let timer = metrics.kv_latency.start_timer();
+        let result = load.await;
+        let elapsed_seconds = timer.stop_and_record();
+
+        if elapsed_seconds > threshold_seconds {
```

Comment:
There's an inconsistency between your approach here and for the postgres reader, where I think you compare milliseconds, and quote the millisecond-based time in the trace (and also the format of your trace message is different). I don't mind which convention we use, but let's be consistent.

Reply:
This also seems like the wrong place to be adding the slow query support -- it should be part of `BigtableReader`, analogous to what you have done with `PgReader` in the earlier PR.

Reply:
Probably this is also the wrong place to add the `kv_*` metrics -- they should also go on `BigtableReader`. Otherwise you are measuring different things for the KV metrics and the DB metrics, and you are missing visibility on how long each individual request to Bigtable is taking (the instrumentation is currently measuring how long a request for a single object via the KV data loader will take, but this is very different, and will be captured by the request-level metrics).